Fix character font lock

rvlo opened the pull request:
Characters escaped weren't being formatted correctly, as shown by the submitted test.

Example of code that wasn't being formatted:
["text" \\, "other"]
(str \\[ "inside" \\])

-----------------

Before submitting a PR mark the checkboxes for the items you've done (if you
think a checkbox does not apply, then leave it unchecked):

- [X] The commits are consistent with our [contribution guidelines][1].
- [X] You've added tests (if possible) to cover your change(s). Bugfix, indentation, and font-lock tests are extremely important!
- [X] You've run `M-x checkdoc` and fixed any warnings in the code you've written.
- [ ] You've updated the changelog (if adding/changing user-visible functionality).
- [ ] You've updated the readme (if adding/changing user-visible functionality).

Thanks!

[1]: https://github.com/clojure-emacs/clojure-mode/blob/master/CONTRIBUTING.md

rvlo commented:
I also propose moving the `lambda arguments` regex below `character literals`, for handling `\%` correctly. Would there be any problems in doing so?

bbatsov commented:
You'll have to check this. The font-locking logic is quite brittle.

The changes you proposed look reasonable. Please, mention the fix in the changelog.

rvlo commented:
I added a test for the % character literal, moved the `lambda arguments` below `character literals`, added an entry to the changelog, rebased on master and squashed the commits.

bbatsov commented:
Thanks!
